### PR TITLE
feat: implement student sorting and filtering functionality

### LINKED ISSUE #123: [FEATURE] : Sort and filter of students based on attendance percentage
# 💡 Feature Description
Sorting and Filter should work 

---

# 🎯 Problem it Solves
This will sort the students which helps to analyze better

---

# 📦 Component Affected
Which part of the system would this feature affect?

- [x] Frontend (UI/UX)

---

# 📸 Mockups/Examples (Optional)

Image

# 📝 Additional Context (Optional)
If any clarity required , comment or DM on linkedin

### PR DESCRIPTION
**Pull request by Omkarop0808**
🔥 Pull Request Summary
This PR implements student sorting and filtering functionality for the Students page as requested in the GitHub issue. Users can now search students by name/roll number, filter by attendance levels (High >90%, Medium 75-90%, Low <75%), and sort by attendance percentage in ascending or descending order.

🔗 Linked Issue
Closes #123  — Implements sorting and filtering functionality for better student data analysis

📦 Type of Change
Check the relevant option(s):

 🐛 Bug fix (non-breaking change which fixes an issue)
 ✨ New feature (non-breaking change which adds functionality)
 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
 📝 Documentation update
 🎨 UI/UX improvement
 ♻️ Code refactoring
 ⚡ Performance improvement
 🧪 Test addition or update

🛠 Changes Made
List the key changes included in this PR:

✅ Search functionality: Real-time search by student name and roll number
✅ Attendance-based filtering: Filter students by High (>90%), Medium (75-90%), Low (<75%) attendance
✅ Sorting by attendance: Toggle between ascending/descending order with visual indicator
✅ Combined filtering: Search + filter + sort work together seamlessly
✅ Dynamic UI updates: Student count updates based on active filters
✅ Empty state handling: Shows appropriate messages when no students match criteria
✅ Interactive UI elements: Filter buttons highlight when selected, sort button shows direction

🧪 Steps to Test
Explain how reviewers can test this PR:

Go to Students page (/students)
Test Search: Type student names or roll numbers in search box
Test Filtering: Click filter buttons (All, High >90%, Medium 75-90%, Low <75%)
Test Sorting: Click "Sort by attendance" button to toggle ascending/descending
Test Combined: Use search + filter + sort together
Test Edge Cases: Search for non-existent names, apply filters with no matches
Verify dynamic count updates in footer ("Showing X of Y students")

🎯 Expected Behaviour
Search filters students in real-time as you type
Filter buttons show only students matching attendance criteria
Sort button toggles between highest-first and lowest-first attendance
Chevron icon rotates when toggling sort direction
Filter buttons highlight when active
Footer shows accurate filtered count
"No students found" message appears when no matches
All functionality works together without conflicts

📸 Screenshots / Proof
Before:
Filter buttons were non-functional (highlighted in red circles in original issue)

After:
Screenshot 2026-02-01 103253

Screenshot 2026-02-01 113148

image

✔ Checklist Before Submitting

 -My code follows the project's code style and conventions
 -Code compiles/runs without errors
 -I have tested my changes locally
 -No breaking changes introduced
 -I have updated relevant documentation (if needed)
 -I have added/updated tests (if applicable)
 -All existing tests are passing
 -I have self-reviewed my code
 -I have commented my code where necessary

📝 Additional Notes (Optional)
Implementation approach: Used React state management with filtering and sorting logic
Performance: Efficient filtering using array methods, no unnecessary re-renders
Backward compatibility: All existing functionality preserved
Code quality: Clean, readable code with proper variable naming
Testing: Comprehensive terminal testing proves all functionality works correctly

## Summary by CodeRabbit

* **New Features**
  * Added search to filter students by name or roll number
  * Sortable attendance percentage with ascending/descending toggle
  * Attendance band filtering (All, High >90%, Medium 75–90%, Low <75%)
  * Improved empty state showing "No students found"
  * Pagination updated to reflect filtered vs total counts
  * Visual attendance indicators with colored badges and progress bars

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: frontend/src/pages/StudentList.jsx [hunk 1]
```diff
@@ -146,11 +202,20 @@ export default function StudentList() {
...
+                      <td colSpan="5" className="px-6 py-12 text-center">
+                        <div className="text-gray-500">
+                          <p className="text-lg font-medium">No students found</p>
+                          <p className="text-sm mt-1">
+                            {searchTerm ? `No students match "${searchTerm}"` : `No students match the selected filter "${selectedFilter}"`}
+                          </p>
+                        </div>
+                      </td>
+                    </tr>
+                  ) : (
```
[the remark this answers is not in the document]
Comment (Omkarop0808) [reply frontend/src/pages/StudentList.jsx]: Thanks for the bot review! I see the suggestion about the empty state message. Should I:
1. Fix the empty state logic now, or  
2. Address it in a follow-up commit?
Happy to make the adjustment
